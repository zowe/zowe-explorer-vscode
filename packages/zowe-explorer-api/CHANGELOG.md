# Change Log

All notable changes to the "zowe-explorer-api" extension will be documented in this file.

## TBD Release

### New features and enhancements

### Bug fixes

- Fixed an issue where the `ZoweVsCodeExtension.updateCredentials` method could remove credentials from session when input prompt was cancelled. [#3018](https://github.com/zowe/zowe-explorer-vscode/pull/3018)
- Updated the `@zowe/cli` dependency to fix error when using session with auth type "none". [zowe-cli#2219](https://github.com/zowe/zowe-cli/issues/2219)

## `2.17.0`

### New features and enhancements

- Updated the `@zowe/cli` dependency to 7.27.0 to support proxy environment variables [#3003](https://github.com/zowe/zowe-explorer-vscode/issues/3003)

### Bug fixes

- Removed `handlebars` dependency in favor of `mustache` for technical currency purposes. [#2974](https://github.com/zowe/zowe-explorer-vscode/pull/2974)

## `2.16.3`

## `2.16.2`

### Bug fixes

- Fixed an issue where the `onProfilesUpdate` event did not fire after secure credentials were updated. [#2822](https://github.com/zowe/zowe-explorer-vscode/issues/2822)
- Update dependencies for technical currency purposes.

## `2.16.1`

## `2.16.0`

### New features and enhancements

- Added optional `consoleName` argument to `ZosmfCommandApi.issueMvsCommand`. [#1667](https://github.com/zowe/vscode-extension-for-zowe/issues/1667)
- Added "Date Completed" attribute to JobSortOpts enum type. [#1685](https://github.com/zowe/vscode-extension-for-zowe/issues/1685)
- Added PEM certificate support as an authentication method for logging into the API ML. [#2621](https://github.com/zowe/zowe-explorer-vscode/issues/2621)

### Bug fixes

- Updated `@zowe/cli` dependency to fix issue where "Log out of authentication service" doesn't show in Manage Profile menu. [#2633](https://github.com/zowe/zowe-explorer-vscode/issues/2633)
- Fixed regression of issue where the `ProfilesCache` class would retain old service profiles, even if they were removed from the team config. [#2910](https://github.com/zowe/zowe-explorer-vscode/issues/2910)

## `2.15.4`

## `2.15.3`

### Bug fixes

- Fixed an issue where `ProfilesCache` may return missing or incorrect profile values when multiple extensions call it during activation. [#2831](https://github.com/zowe/zowe-explorer-vscode/issues/2831)

## `2.15.2`

### New features and enhancements

- Added optional method `getDsDocumentFilePath` to `IZoweDatasetTreeNode` interface to make it easier for extenders to get the local file path of a data set node. [#2760](https://github.com/zowe/vscode-extension-for-zowe/pull/2760)

### Bug fixes

- Fixed an issue where the `ProfilesCache` class would retain old service profiles, even if they were removed from the team config. [#2395](https://github.com/zowe/zowe-explorer-vscode/issues/2395)

## `2.15.1`

### Bug fixes

- Fixed TypeError encountered in the `ProfilesCache.checkMergingConfigAllProfiles` function when merging profiles. [#2771](https://github.com/zowe/vscode-extension-for-zowe/pull/2771)

## `2.15.0`

### Bug fixes

- Fix login and logout operations when APIML dynamic tokens are enabled. [#2692](https://github.com/zowe/vscode-extension-for-zowe/pull/2692)
- Fixed issue where `zosmf` profiles did not respect the `protocol` property. [#2703](https://github.com/zowe/vscode-extension-for-zowe/issues/2703)
- Fix to restore accessibility to all profiles when default profile has APIML token authentication. [#2111](https://github.com/zowe/vscode-extension-for-zowe/issues/2111)

## `2.14.1`

### Bug fixes

- Update transitive dependencies for technical currency.

## `2.14.0`

### New features and enhancements

- Added optional `openDs` function to `IZoweDatasetTreeNode` to open a data set or member in the editor.
- Added optional `setEncoding` function to `IZoweDatasetTreeNode` and `IZoweUSSTreeNode` to set the encoding of a node to binary, text, or a custom codepage.
- Added optional properties `binary`, `encoding`, and `encodingMap` to tree node interfaces for storing the codepage of a data set or USS file.
- Deprecated `IZoweUSSTreeNode.binaryFiles` and `IZoweUSSTreeNode.setBinary` in favor of `IZoweUSSTreeNode.encodingMap` and `IZoweUSSTreeNode.setEncoding`.
- Deprecated `ZoweTreeNode.binary`, `ZoweTreeNode.binaryFiles`, and `ZoweTreeNode.shortLabel`. These properties are not applicable for all tree nodes and should be defined in subclasses of `ZoweTreeNode` if necessary.
- Added new functions `loginWithBaseProfile` and `logoutWithBaseProfile` to provide extenders with the ability to automatically login to their respective services. [#2493](https://github.com/zowe/vscode-extension-for-zowe/pull/2493)
- Added APIML dynamic token support. [#2665](https://github.com/zowe/vscode-extension-for-zowe/issues/2665)
- Added new optional method `getCommonApi` to `ZoweExplorerApi.IApiRegisterClient` for enhanced typings in other Zowe Explorer APIs. [#2493](https://github.com/zowe/vscode-extension-for-zowe/pull/2493)
- Add Created Date to `stats` optional variable for storing dataset stats
- Add Date created to DatasetSortOpts enum [#2707](https://github.com/zowe/vscode-extension-for-zowe/pull/2707)

### Bug fixes

- Added a return type of void for `IZoweUSSTreeNode.openUSS`.
- Fixed use of `this` in static methods in `ZoweVsCodeExtension`. [#2606](https://github.com/zowe/vscode-extension-for-zowe/pull/2606)
- Fixed `ZoweVsCodeExtension.promptUserPass` to not use hardcoded values for user and password. [#2666](https://github.com/zowe/vscode-extension-for-zowe/issues/2666)

## `2.13.1`

### Bug fixes

- Update dependencies for technical currency purposes.

## `2.13.0`

### New features and enhancements

- Added new optional boolean parameter `hideFromAllTrees` to `IZoweTree.deleteSession` for specifying whether to hide from all trees or current tree. [#2567](https://github.com/zowe/vscode-extension-for-zowe/issues/2567)
- Added new optional parameter `provider` of type `IZoweTree<IZoweTreeNode>` for `IZoweTree.addSession` to specify a tree to add the profile to.
- Added optional `filter` and `actualJobs` variables to `IZoweJobTreeNode` to track local filter search.
- Added new optional record `openFiles` to `IZoweTree` to track opened files under a specific tree view. [#2597](https://github.com/zowe/vscode-extension-for-zowe/issues/2597)

## `2.12.2`

## `2.12.1`

## `2.12.0`

### New features and enhancements

- Added optional `getTag` function to `ZoweExplorerAPI.IUss` for getting the tag of a file on USS.
- Added new ProfilesCache API `getConfigArray` to allow extenders to get the registered profile type's metadata for team configuration file creation handled outside of Zowe Explorer views. [#2394](https://github.com/zowe/vscode-extension-for-zowe/issues/2394)
- Add `sort` and `filter` optional variables for storing sort/filter options alongside tree nodes. [#2420](https://github.com/zowe/vscode-extension-for-zowe/issues/2420)
- Add `stats` optional variable for storing dataset stats (such as user, modified date, etc.).
- Add option enums and types for sorting, filtering and sort direction in tree nodes. [#2420](https://github.com/zowe/vscode-extension-for-zowe/issues/2420)
- Added option for retaining context when generating webviews in Webview API.

## `2.11.2`

### Bug fixes

- Bundle Zowe Secrets for issues seen by extenders that use the ProfilesCache for profile management. [#2512](https://github.com/zowe/vscode-extension-for-zowe/issues/2512)

## `2.11.1`

## `2.11.0`

### New features and enhancements

- Added optional `pendingActions` record to `IZoweTreeNode` to allow nodes to track pending promises.
- Added optional `wasDoubleClicked` variable to `IZoweTreeNode` to track whether a node was double-clicked during an action.

### Bug fixes

- Bump `@zowe/secrets-for-zowe-sdk` to 7.18.4 to handle install errors gracefully and to allow running without MSVC redistributables.

## `2.10.0`

### New features and enhancements

- Added option to register callback to be called after making changes to team config profiles are made. [#2385](https://github.com/zowe/vscode-extension-for-zowe/issues/2385)
- Replaced `keytar` dependency with `keyring` module from [`@zowe/secrets-for-zowe-sdk`](https://github.com/zowe/zowe-cli/tree/master/packages/secrets). [#2358](https://github.com/zowe/vscode-extension-for-zowe/issues/2358) [#2348](https://github.com/zowe/vscode-extension-for-zowe/issues/2348)
- Added `WebView` class to allow Zowe Explorer and extenders to create enhanced webviews (choose any JavaScript bundler and JavaScript framework). [#2254](https://github.com/zowe/vscode-extension-for-zowe/issues/2254)

## `2.9.2`

## `2.9.1`

### New features and enhancements

- Added optional `profile` parameter to `IPromptCredentialsOptions` so developers can choose to skip rebuilding the profile with ProfilesCache.

### Bug fixes

- Fixed error when an extender's extension attempts to access the keyring in a remote VSCode session [#324](https://github.com/zowe/vscode-extension-for-cics/issues/324).
- Fixed issue where profiles with authentication tokens were breaking functionality for direct-to-service profiles after user interaction. [#2330](https://github.com/zowe/vscode-extension-for-zowe/issues/2330)
- Updated dependencies for security audits.

## `2.9.0`

### New features and enhancements

- Added optional IZoweTree functions, `addDsTemplate` and `getDSTemplates`.
- Added a new type `DataSetAllocTemplate` for data set attributes.
- Added optional `cancelJob` function to `ZoweExplorerApi.IJes` interface.
- Added z/OSMF API implementation for `cancelJob` function.
- Added optional `id` variable to `IZoweTreeNode` interface, which can be used to designate a unique ID for a tree node. [#2215](https://github.com/zowe/vscode-extension-for-zowe/issues/2215)
- Fixed error shown by API when accessing the `name` and `type` property of a profile when updating the profile arrays [#2334](https://github.com/zowe/vscode-extension-for-zowe/issues/2334).

## `2.8.1`

### Bug fixes

- Updated linter rules and addressed linter errors. [#2291](https://github.com/zowe/vscode-extension-for-zowe/issues/2291)
- Updated dependencies for security audits.

## `2.8.0`

### New features and enhancements

- Added `Gui.reportProgress` that can be used to notify users of action progress in conjunction with the `Gui.withProgress` call. [#2167](https://github.com/zowe/vscode-extension-for-zowe/issues/2167)
- Updated linter rules and addressed linter errors. [#2184](https://github.com/zowe/vscode-extension-for-zowe/issues/2184)
- Added checks to verify that `@zowe/cli` dependency exists before building. [#2199](https://github.com/zowe/vscode-extension-for-zowe/issues/2199)
- Added `ZoweVsCodeExtension.customLoggingPath` that can be used to get custom logging path defined in VS Code settings. [#2186](https://github.com/zowe/vscode-extension-for-zowe/issues/2186)
- Added `Poller` utility singleton for handling continuous poll requests: see `Poller.addRequest, Poller.removeRequest` functions.
- Added `pollData` optional function to `IZoweTree` class.
- Created a new optional API, `IJes.downloadSingleSpool`, that can be used to download a single spool file in text or binary formats. [#2060](https://github.com/zowe/vscode-extension-for-zowe/issues/2060)
- Added capability to download a single spool file to in text or binary formats for zOSMF profiles by adopting the new api, `ZosmfJesApi.downloadSingleSpool`. [#2060](https://github.com/zowe/vscode-extension-for-zowe/issues/2060)

### Bug fixes

- Fixed credentials being updated for wrong v1 profile if multiple profiles had different types but same name.
- Updated dependencies for security audits.
- Added fallback for `realPathSync` to resolve edge cases where the native call fails on Windows systems. [#1773](https://github.com/zowe/vscode-extension-for-zowe/issues/1773)

## `2.7.0`

### New features and enhancements

- Updated `IZoweTreeNode` with additional variable `description` to prevent compilation errors when updating node descriptions. [#2122](https://github.com/zowe/vscode-extension-for-zowe/issues/2122)
- Updated `IZoweJobTreeNode` with additional variable `filtered` to track whether a job session node has been filtered. [#2122](https://github.com/zowe/vscode-extension-for-zowe/issues/2122)
- Added new API `IMvs.copyDataSet`. [#1550](https://github.com/zowe/vscode-extension-for-zowe/issues/1550)

## `2.6.2`

### Bug fixes

- Updated dependencies for security audits.

## `2.6.0`

### New features and enhancements

- Refactored UI/UX methods into standalone `Gui` module for usability and maintainability. [#1967](https://github.com/zowe/vscode-extension-for-zowe/issues/1967)
- New API call `dataSetsMatchingPattern` to allow filtering datasets via a pattern.
- Added `copy` function to USS API to facilitate with copying files.

### Bug fixes

- Updated Imperative to fix failure to load schema when there is no profile of that type. [zowe/imperative#916](https://github.com/zowe/imperative/pull/916)
- Added missing overload for `Gui.setStatusBarMessage` to allow passing `Thenable` objects.

## `2.5.0`

- Copy and Paste added to IZoweTree API for files and directories on USS tree.

## `2.4.1`

### Bug fixes

- Added logging in places where errors were being caught and ignored.
- Fixed all existing ESLint errors within the API logic.
- Removed TSLint (as it is deprecated), and replaced all TSLint rules with their ESLint equivalents. [#2030](https://github.com/zowe/vscode-extension-for-zowe/issues/2030)

### New features and enhancements

- New API call `getJobsByParameters` to allow filtering jobs by status.
- Added `findEquivalentNode` function to IZoweTree to find a corresponding favorited/non-favorited node.
- Updated `IZoweTree`: changed `IZoweNodeType -> IZoweTreeNode` to prevent incompatibility w/ custom/future Zowe node types

## `2.4.1`

### Bug fixes

- Added an API to obtain an up to date array of Profiles from registered types, `ProfilesCache.fetchAllProfiles()`.
- Fixed `ZoweVsCodeExtension` failing to initialize in environment with empty workspace. [#1994](https://github.com/zowe/vscode-extension-for-zowe/issues/1994)

## `2.4.0`

- Fixed refresh for Zowe Explorer activation and Refresh Extension issues in web based editors. [#1807](https://github.com/zowe/vscode-extension-for-zowe/issues/1807)

## `2.2.1`

- Bugfix: Fix for extenders that call registerCustomProfileType() and recieved error when team configuration file was in place. [#1870](https://github.com/zowe/vscode-extension-for-zowe/issues/1870)

## `2.2.0`

- New API `ZoweVsCodeExtension.updateCredentials` for credential prompting that updates the ProfilesCache after obtaining credentials from user.
- New API `ProfilesCache.updateProfilesArrays` to update `ProfilesCache.allProfiles` for profiles that don't store credentials locally in profile file.
- New API `ProfilesCache.isCredentialsSecured` to check if credentials are stored securely.
- Deprecated `ZoweVsCodeExtension.promptCredentials` in favor of `ZoweVsCodeExtension.updateCredentials`.

## `2.0.0`

- Major: Introduced Team Profiles and more. See the prerelease items (if any) below for more details.

## `2.0.0-next.202204081040`

- Added documentation on promptCredentials(). [1728](https://github.com/zowe/vscode-extension-for-zowe/pull/1728)
- Updated ProfilesCache.refresh() to handle the merging on v1 and v2 profiles. [1729](https://github.com/zowe/vscode-extension-for-zowe/pull/1729)

## `2.0.0-next.202204041200`

- Added new API to expose `promptCredentials` for extender use. [#1699](https://github.com/zowe/vscode-extension-for-zowe/pull/1699)

## `1.17.0`

- Zowe Explorer extenders can now have their profile type's folder with meta file created in the /.zowe/profiles home directory upon initialization by calling the ZoweExplorerApiRegister.getExplorerExtenderApi().initForZowe(type: string, meta:imperative.ICommandProfileTypeConfiguration[]) during their activation with Zowe Explorer.

## `1.10.1`

- Initial release
