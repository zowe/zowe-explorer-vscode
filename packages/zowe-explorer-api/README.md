# Extensibility API for Zowe Explorer

[![version](https://img.shields.io/npm/v/@zowe/zowe-explorer-api)](https://img.shields.io/npm/v/@zowe/zowe-explorer-api)
[![downloads](https://img.shields.io/npm/dt/@zowe/zowe-explorer-api)](https://img.shields.io/npm/dt/@zowe/zowe-explorer-api)

Extensibility API for Zowe Explorer is a collection of APIs that can be used to extend the [Zowe Explorer](https://github.com/zowe/zowe-explorer-vscode) VS Code extension with alternative z/OS interaction protocols and new capabilities.

The API is used by extenders of the Zowe Explorer VS Code extension, including the [IBM z/OS FTP for Zowe Explorer](https://github.com/zowe/zowe-explorer-vscode/tree/main/packages/zowe-explorer-ftp-extension) and [IBM CICS for Zowe Explorer](https://github.com/zowe/cics-for-zowe-client/tree/main/packages/vsce) VS Code extensions, as well as several commercial extensions.

The API is organized into modules that are exposed through the top-level `index.ts` file for convenient access.

- `/profiles`: Provides access to common Zowe CLI profile management APIs, as well as abstractions for providing alternative z/OS interactions that use protocols other than z/OSMF, based on alternative Zowe CLI profile types.
- `/tree`: Provides abstractions for accessing and extending the Zowe Explorer VS Code tree views.
- `/logger`: Logs messages in a standard format, which is used by Zowe Explorer and will be consistent across Zowe components.

## Profiles API

The `/profiles` module has no dependency to VS Code and could be used in Zowe CLI extensions as well as VS Code extensions. If you want to use it without pulling in dependencies to VS Code that are needed by the `/tree` module, you can import it like this:

```ts
import { ZoweExplorerApi } from "@zowe/zowe-explorer-api/lib/profiles";
```

The main API provided by the `/profiles` module is called `ZoweExplorerApi`. It defines a namespace for interfaces for implementing access to z/OS MVS, USS, and JES. You can see that the Zowe Explorer FTP Extension provides such an alternative implementation for USS using the FTP protocol in [packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpMvsApi.ts](https://github.com/zowe/zowe-explorer-vscode/tree/main/packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpMvsApi.ts). The `/profiles` module itself contains Zowe Explorer's default implementation using z/OSMF in [packages/zowe-explorer-api/src/profiles/ZoweExplorerZosmfApi.ts](https://github.com/zowe/zowe-explorer-vscode/tree/main/packages/zowe-explorer-api/src/profiles/ZoweExplorerZosmfApi.ts)

Zowe Explorer itself exports a `IApiRegisterClient` object that can be used for an alternative implementation to be registered with Zowe Explorer. You can find an example for doing this in [packages/zowe-explorer-ftp-extension/src/extension.ts](https://github.com/zowe/zowe-explorer-vscode/tree/main/packages/zowe-explorer-ftp-extension/src/extension.ts). To be able to do this, your VS Code extension must define a dependency to Zowe Explorer to ensure that VS Code extension activation is performed in the correct order. Therefore, your VS Code extension's `package.json` must contain

```json
"extensionDependencies": [
  "Zowe.vscode-extension-for-zowe"
]
```

## Tree API

Importing from the `/tree` module requires that your NPM package has a dependency to VS Code as it performs

```ts
import * as vscode from "vscode";
```

imports.

See this [documentation on Extending Zowe Explorer](https://github.com/zowe/zowe-explorer-vscode/wiki/Extending-Zowe-Explorer) to learn more about the Tree APIs available.

## Logger API

See this [special extension document](https://github.com/zowe/zowe-explorer-vscode/wiki/Error-Handling-for-Extenders#logging-of-error-message) for more information about using the Logger API.

## Providing feedback or help contributing

Extensibility API for Zowe Explorer is part of the [Zowe Explorer monorepo on Github](https://github.com/zowe/zowe-explorer-vscode). You find the sources there in the `/packages/zowe-explorer-api` sub-folder.

To submit a bug report or enhancement request, open an issue in the [Zowe Explorer repository](https://github.com/zowe/zowe-explorer-vscode/issues).
