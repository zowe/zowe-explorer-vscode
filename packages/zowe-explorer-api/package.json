{
  "name": "@zowe/zowe-explorer-api",
  "version": "3.1.0-SNAPSHOT",
  "description": "Extensibility API for Zowe Explorer.",
  "publisher": "Zowe",
  "author": "Zowe",
  "license": "EPL-2.0",
  "repository": {
    "url": "https://github.com/zowe/zowe-explorer-vscode"
  },
  "publishConfig": {
    "provenance": true
  },
  "main": "lib/index.js",
  "exports": {
    ".": {
      "@zowe:bundler": "./src/index.ts",
      "default": "./lib/index.js"
    }
  },
  "files": [
    "lib"
  ],
  "devDependencies": {
    "@types/mustache": "^4.2.5",
    "@types/semver": "^7.5.0",
    "copyfiles": "^2.4.1"
  },
  "dependencies": {
    "@types/vscode": "^1.53.2",
    "@zowe/core-for-zowe-sdk": "^8.1.1",
    "@zowe/imperative": "^8.1.0",
    "@zowe/secrets-for-zowe-sdk": "^8.1.0",
    "@zowe/zos-console-for-zowe-sdk": "^8.1.1",
    "@zowe/zos-files-for-zowe-sdk": "^8.1.1",
    "@zowe/zos-jobs-for-zowe-sdk": "^8.1.1",
    "@zowe/zos-tso-for-zowe-sdk": "^8.1.1",
    "@zowe/zos-uss-for-zowe-sdk": "^8.1.1",
    "@zowe/zosmf-for-zowe-sdk": "^8.1.1",
    "deep-object-diff": "^1.1.9",
    "mustache": "^4.2.0",
    "semver": "^7.6.0"
  },
  "scripts": {
    "build": "pnpm check-cli && pnpm copy-secrets && pnpm clean && pnpm license && tsc -p ./ && pnpm madge",
    "test:unit": "jest \".*__tests__.*\\.unit\\.test\\.ts\" --coverage",
    "test": "pnpm test:unit",
    "lint": "concurrently -n \"_eslint_,prettier\" \"pnpm lint:all\" \"pnpm lint:pretty\"",
    "lint:all": "pnpm run \"/^lint:[hst].*/\"",
    "lint:html": "(mkdir -p \"results\" || true) && eslint . --format html > results/eslint.html",
    "lint:pretty": "prettier --write . && prettier --check .",
    "lint:src": "eslint --format stylish src/**/*.ts",
    "lint:test": "eslint --format stylish __tests__/**/*.ts",
    "madge": "madge -c --no-color --no-spinner --exclude __mocks__ --extensions js,ts src/",
    "pretty": "prettier --write .",
    "pretty:check": "prettier --check .",
    "check-cli": "node scripts/check-cli.js",
    "clean": "rimraf lib tsconfig.tsbuildinfo",
    "fresh-clone": "pnpm clean && (rimraf node_modules || true)",
    "license": "node ../../scripts/updateLicenses.js",
    "package": "pnpm build && pnpm pack && node ../../scripts/mv-pack.js zowe-zowe-explorer-api tgz",
    "copy-secrets": "copyfiles -f ../../node_modules/@zowe/secrets-for-zowe-sdk/prebuilds/*.node ./prebuilds"
  },
  "jest": {
    "moduleFileExtensions": [
      "js",
      "ts",
      "tsx"
    ],
    "testPathIgnorePatterns": [
      "<rootDir>/src/decorators"
    ],
    "watchPathIgnorePatterns": [
      "<rootDir>/results/unit"
    ],
    "transform": {
      "^.+\\.tsx?$": [
        "ts-jest",
        {
          "diagnostics": false,
          "tsconfig": "tsconfig-tests.json"
        }
      ]
    },
    "testRegex": "__tests__.*\\.(spec|test)\\.ts$",
    "modulePathIgnorePatterns": [
      "out/"
    ],
    "reporters": [
      "default",
      "jest-junit",
      "jest-stare",
      "jest-html-reporter"
    ],
    "coverageReporters": [
      "json",
      "lcov",
      "text",
      "cobertura"
    ],
    "coverageDirectory": "<rootDir>/results/unit/coverage",
    "preset": "ts-jest"
  },
  "jest-html-reporter": {
    "pageTitle": "Zowe Explorer API Tests",
    "outputPath": "results/unit/results.html"
  },
  "jest-stare": {
    "resultDir": "results/unit/jest-stare",
    "coverageLink": "../coverage/lcov-report/index.html"
  },
  "jest-junit": {
    "outputDirectory": "<rootDir>/results/unit",
    "outputName": "junit.xml"
  },
  "madge": {
    "detectiveOptions": {
      "ts": {
        "skipTypeImports": true
      }
    }
  }
}
