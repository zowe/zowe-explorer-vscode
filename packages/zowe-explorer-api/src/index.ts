/**
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

export * from "./dataset";
export * from "./extend";
export * from "./globals/Gui";
export * from "./globals/GuiOptions";
export * from "./logger";
export * from "./profiles";
export * from "./security";
export * from "./tree";
export * from "./utils";
export * from "./vscode";
export * from "./Types";
export * from "./fs";
export * as imperative from "@zowe/imperative";
