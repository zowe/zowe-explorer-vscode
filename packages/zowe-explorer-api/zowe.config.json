{
  "profiles": {
    "lpar": {
      "profiles": {
        "service": {
          "type": "service",
          "properties": []
        }
      },
      "properties": []
    }
  },
  "defaults": {}
}
