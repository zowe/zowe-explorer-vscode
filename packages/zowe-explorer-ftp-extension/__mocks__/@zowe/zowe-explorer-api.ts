/**
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

export * as imperative from "@zowe/imperative";

// This shouldn't be considered redeclared, as its contained within a mock.
export class MessageSeverity {
    public static TRACE = 0;
    public static DEBUG = 1;
    public static INFO = 2;
    public static WARN = 3;
    public static ERROR = 4;
    public static FATAL = 5;
}

export class IZoweLogger {
    public logImperativeMessage(message: string, severity: MessageSeverity): void {
        return undefined;
    }
}

export interface GuiMessageOptions {
    severity?: MessageSeverity;
    items?: string[];
    logger?: IZoweLogger;
    vsCodeOpts?: object;
}

export namespace Gui {
    /**
     * Displays a message to the user, including any specified options.
     * @param message The message to display
     * @param options Any additional options for the message
     * @returns A thenable containing the selected item (if items were specified), or `undefined`
     */
    export function showMessage(message: string, options?: GuiMessageOptions): Thenable<string> {
        return undefined;
    }

    export function errorMessage(message: string, options?: Omit<GuiMessageOptions, "severity">): Thenable<string> {
        return undefined;
    }

    export function warningMessage() {
        return undefined;
    }
}

export class ZoweVsCodeExtension {
    public static get customLoggingPath(): string | undefined {
        return undefined;
    }

    public static getZoweExplorerApi(requiredVersion?: string): any {
        return {
            registerUssApi: () => {},
            registerJesApi: () => {},
            registerMvsApi: () => {},
            getExplorerExtenderApi: () => ({
                initForZowe: () => {},
                reloadProfiles: () => {},
            }),
        };
    }
}

export class AuthHandler {
    public static enableLocksForType(typ: string): void {
        return;
    }
}
