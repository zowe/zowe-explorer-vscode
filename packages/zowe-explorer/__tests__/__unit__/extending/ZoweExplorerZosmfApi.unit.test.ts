/**
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import * as zosfiles from "@zowe/zos-files-for-zowe-sdk";
import { imperative, ZoweExplorerZosmf } from "@zowe/zowe-explorer-api";

export declare enum TaskStage {
    IN_PROGRESS = 0,
    COMPLETE = 1,
    NOT_STARTED = 2,
    FAILED = 3,
}

describe("Zosmf API tests", () => {
    it("should test that copy data set uses default options", async () => {
        const dataSet = jest.fn((_session, _toDataSet, options) => {
            expect(options).toMatchSnapshot();
            return { api: "", commandResponse: "", success: true };
        });

        (zosfiles as any).Copy = { dataSet };

        const api = new ZoweExplorerZosmf.MvsApi();
        api.getSession = jest.fn();
        await api.copyDataSetMember({ dsn: "IBM.FROM", member: "IEFBR14" }, { dsn: "IBM.TO", member: "IEFBR15" }, {
            responseTimeout: undefined,
        } as any);
    });

    it("should test that copy data set uses enq", async () => {
        const dataSet = jest.fn((_session, _toDataSet, options) => {
            expect(options).toMatchSnapshot();
            return { api: "", commandResponse: "", success: true };
        });

        (zosfiles as any).Copy = { dataSet };

        const api = new ZoweExplorerZosmf.MvsApi();
        api.getSession = jest.fn();
        await api.copyDataSetMember(
            { dsn: "IBM.FROM", member: "IEFBR14" },
            { dsn: "IBM.TO", member: "IEFBR15" },
            { enq: "SHR", "from-dataset": { dsn: "BROADCOM.FROM" }, responseTimeout: undefined }
        );
    });

    it("should test that copy data set uses enq only", async () => {
        const dataSet = jest.fn((_session, _toDataSet, options) => {
            expect(options).toMatchSnapshot();
            return { api: "", commandResponse: "", success: true };
        });

        (zosfiles as any).Copy = { dataSet };

        const api = new ZoweExplorerZosmf.MvsApi();
        api.getSession = jest.fn();
        await api.copyDataSetMember({ dsn: "IBM.FROM", member: "IEFBR14" }, { dsn: "IBM.TO", member: "IEFBR15" }, {
            enq: "SHR",
            responseTimeout: undefined,
        } as any);
    });

    it("should test putContent method passes all options to Zowe api method", async () => {
        const fileToUssFile = jest.fn(
            (_session: imperative.AbstractSession, _inputFile: string, _ussname: string, options?: zosfiles.IUploadOptions) => {
                expect(options).toMatchSnapshot();
                return { api: "", commandResponse: "", success: true };
            }
        );

        (zosfiles as any).Upload = { fileToUssFile };

        const api = new ZoweExplorerZosmf.UssApi();
        api.getSession = jest.fn();

        await api.putContent("someLocalFile.txt", "/some/remote", {
            encoding: "285",
            responseTimeout: undefined,
        });
    });

    it("should test that getContents calls zowe.Download.ussFile", async () => {
        const api = new ZoweExplorerZosmf.UssApi();
        api.getSession = jest.fn();
        const response = { shouldMatch: true };

        Object.defineProperty(zosfiles, "Download", {
            value: {
                ussFile: jest.fn().mockResolvedValue(response),
            },
            configurable: true,
        });

        await expect(api.getContents("/some/input/path", {})).resolves.toEqual(response);
    });

    it("should update the tag attribute of a USS file if a new change is made", async () => {
        const api = new ZoweExplorerZosmf.UssApi();
        const changeTagSpy = jest.fn();
        Object.defineProperty(zosfiles, "Utilities", {
            value: {
                putUSSPayload: changeTagSpy,
            },
            configurable: true,
        });
        await expect(api.updateAttributes("/test/path", { tag: "utf-8" })).resolves.not.toThrow();
        expect(changeTagSpy).toHaveBeenCalledTimes(1);
    });

    it("should get the tag of a file successfully", async () => {
        const api = new ZoweExplorerZosmf.UssApi();
        jest.spyOn(JSON, "parse").mockReturnValue({
            stdout: ["-t UTF-8 tesfile.txt"],
        });
        Object.defineProperty(zosfiles, "Utilities", {
            value: {
                putUSSPayload: () => Buffer.from(""),
            },
            configurable: true,
        });
        await expect(api.getTag("testfile.txt")).resolves.toEqual("UTF-8");
    });
});
