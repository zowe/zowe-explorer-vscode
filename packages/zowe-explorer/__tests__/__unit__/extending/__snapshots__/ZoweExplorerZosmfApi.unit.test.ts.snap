// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Zosmf API tests should test putContent method passes all options to Zowe api method 1`] = `
{
  "encoding": "285",
  "responseTimeout": undefined,
}
`;

exports[`Zosmf API tests should test that copy data set uses default options 1`] = `
{
  "from-dataset": {
    "dsn": "IBM.FROM",
    "member": "IEFBR14",
  },
  "responseTimeout": undefined,
}
`;

exports[`Zosmf API tests should test that copy data set uses enq 1`] = `
{
  "enq": "SHR",
  "from-dataset": {
    "dsn": "BROADCOM.FROM",
  },
  "responseTimeout": undefined,
}
`;

exports[`Zosmf API tests should test that copy data set uses enq only 1`] = `
{
  "enq": "SHR",
  "from-dataset": {
    "dsn": "IBM.FROM",
    "member": "IEFBR14",
  },
  "responseTimeout": undefined,
}
`;
